### PR TITLE
Feature 111 form editor listing

### LINKED ISSUE #111: Editor list view
The screen that shows all available (to current user) Metaforms and their MetaformVersions is not yet implemented.
Buttons to create new Metaforms and edit existing ones are located in this screen and therefore it should be implemented.

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #111.
Contains screen for admin form editor listing.
Does not yet handle conversion of Sosmeta form schemas.

### CODE REVIEW

File: src/components/screens/editor/editor-screen.tsx [hunk 1]
```diff
@@ -1,12 +1,273 @@
...
+        <Divider/>
+      </Grid>
+    );
+  };
+
+  /**
+   * Toggles drawer
+   */
+  const toggleEditorDrawer = () => {
+    setDrawerOpen(!drawerOpen);
```
Comment (WiljamiT) [inline src/components/screens/editor/editor-screen.tsx:173]: setDrawerOpen((prev: boolean) => !prev); or setDrawerOpen((prevState: boolean) => !prevState);

Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:173]: Done.

**Review by WiljamiT — APPROVED** [review]

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+  /**
+   * Renders Drawer identification section
+   */
+  const renderDrawerIdentificationSection = () => {
+    return (
+      <Box
+        sx={{ padding: 2 }}
+      >
+        <Stack spacing={2}>
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:212]:
```suggestion
        <Stack spacing={ 2 }>
```

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+          />
+        </Stack>
+      </Box>
+    );
+  };
+
+  /**
+   * Renders Drawer identification section
+   */
+  const renderDrawerIdentificationSection = () => {
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:207]: Identification = Authentication

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+          </RadioGroup>
+        </Stack>
+      </Box>
+    );
+  };
+
+  useEffect(() => {
+    const templateSectionValid = formSettings.formTemplate ? formSettings.formTemplate && !!formSettings.formSchema : !formSettings.formTemplate;
+    setValid(!!formSettings.formName && templateSectionValid);
+  }, [formSettings]);
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:234]: For readability, consider making methods that you run inside useEffect(), don't write logic inside it.

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+  useEffect(() => {
+    if (!formSettings.formTemplate) {
+      setFormSettings({ ...formSettings, formSchema: "" });
+    }
+  }, [formSettings.formTemplate]);
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:240]: Same here

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+      PaperProps={{
+        sx: {
+          width: "400px", borderRadius: "15px 0 0 15px"
+        }
+      }}
+      anchor="right"
+      open={ open }
+      onClose={ toggleDrawerOpen }
+    >
+      <Stack spacing={2} direction="column">
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:253]:
```suggestion
      <Stack spacing={ 2 } direction="column">
```

File: src/localization/fi.json [hunk 7]
```diff
@@ -252,6 +257,33 @@
     "noTitle": "Ei otsikkoa"
   },
 
+  "editorScreen": {
+    "noMetaforms": "Metaformeja ei löytynyt",
```
Comment (EVP1989) [inline src/localization/fi.json:261]:
```suggestion
    "noMetaforms": "Lomakkeita ei löytynyt",
```

File: src/localization/en.json [hunk 8]
```diff
@@ -253,6 +258,33 @@
     "noTitle": "No title"
   },
 
+  "editorScreen": {
+    "noMetaforms": "No Metaforms found",
```
Comment (EVP1989) [inline src/localization/en.json:262]:
```suggestion
    "noMetaforms": "No forms found",
```

File: src/components/screens/editor/editor-screen.tsx [hunk 9]
```diff
@@ -1,12 +1,273 @@
...
+        <Divider/>
+      </Grid>
+    );
+  };
+
+  /**
+   * Toggles drawer
+   */
+  const toggleEditorDrawer = () => {
+    setDrawerOpen((prevState: boolean) => !prevState);
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:176]:
```suggestion
    setDrawerOpen(!!drawerOpen);
```

File: src/components/screens/editor/editor-screen.tsx [hunk 10]
```diff
@@ -1,12 +1,273 @@
...
+  const buildVersionRows = (metaform: Metaform) => {
+    const versions = metaforms.find(x => x.metaform.id === metaform.id)?.versions;
+    if (!versions) {
+      return;
+    }
+
+    return versions.map((version: MetaformVersion) => {
+      return (
+        <ListItem
+          key={ version.id!! }
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:98]:
```suggestion
          key={ version.id! }
```

Comment (EVP1989) [reply src/components/screens/editor/editor-screen.tsx:98]: In javascript, only one exclamation mark is needed.

File: src/components/screens/editor/editor-screen.tsx [hunk 11]
```diff
@@ -1,12 +1,273 @@
...
+          sx={{ width: "100%" }}
+        >
+          <AdminFormListStack direction="row">
+            <Grid container>
+              <Grid item md={9}>
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
              <Grid item md={ 9 }>
```

File: src/components/screens/editor/editor-screen.tsx [hunk 12]
```diff
@@ -1,12 +1,273 @@
...
+                <AdminFormTypographyField>
+                  <ListRounded style={{ fill: "darkgrey" }}/>
+                  { version.type === MetaformVersionType.Archived ? strings.editorScreen.formVersionArchived : strings.editorScreen.formVersionDraft }
+                </AdminFormTypographyField>
+              </Grid>
+              <Grid item md={1}>
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
              <Grid item md={ 1 }>
```

File: src/components/screens/editor/editor-screen.tsx [hunk 13]
```diff
@@ -1,12 +1,273 @@
...
+                <AdminFormTypographyField>
+                  <DateRangeRounded style={{ fill: "darkgrey" }}/>
+                  { moment(version.createdAt).format("DD.MM.yyyy HH:mm") }
+                </AdminFormTypographyField>
+              </Grid>
+              <Grid item md={1}>
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
              <Grid item md={ 1 }>
```

File: src/components/screens/editor/editor-screen.tsx [hunk 14]
```diff
@@ -1,12 +1,273 @@
...
+                <AdminFormTypographyField>
+                  <DateRangeRounded style={{ fill: "darkgrey" }}/>
+                  { moment(version.modifiedAt).format("DD.MM.yyyy HH:mm") }
+                </AdminFormTypographyField>
+              </Grid>
+              <Grid item md={1}>
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx]:
```suggestion
              <Grid item md={ 1 }>
```

File: src/components/screens/editor/editor-screen.tsx [hunk 15]
```diff
@@ -1,12 +1,273 @@
...
+  };
+
+  /**
+   * Creates new Metaform and MetaformVersion
+   */
+  const createMetaform = async (metaform: Metaform) => {
+    try {
+      const newMetaform = await metaformsApi.createMetaform({ metaform: metaform });
+      const newMetaformVersion = await versionsApi.createMetaformVersion({
+        metaformId: newMetaform.id!!,
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:71]:
```suggestion
        metaformId: newMetaform.id!,
```

File: src/components/screens/editor/editor-screen.tsx [hunk 16]
```diff
@@ -1,12 +1,273 @@
...
+        metaformVersion: {
+          type: MetaformVersionType.Draft,
+          data: { ...newMetaform } as any
+        }
+      });
+      const currentPath = window.location.pathname;
+      navigate(`${currentPath}/${newMetaform.slug}/${newMetaformVersion.id}`);
+    } catch (e: any) {
+      errorContext.setError("Error while creating new Metaform.");
+    }
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:81]: Use localized error message and pass actual error also to error handler, no need to type error as any.

File: src/components/screens/editor/editor-screen.tsx [hunk 17]
```diff
@@ -1,12 +1,273 @@
...
+        return {
+          metaform: form,
+          versions: await versionsApi.listMetaformVersions({ metaformId: form.id!! })
+        };
+      }));
+
+      setMetaforms(rows);
+    } catch (e: any) {
+      errorContext.setError("Error while listing forms.");
+    }
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:59]: Check error handling (localiaztion, error doesn't need typing etc)

File: src/components/screens/editor/editor-screen.tsx [hunk 18]
```diff
@@ -1,12 +1,273 @@
...
+   */
+  const loadMetaforms = async () => {
+    setLoading(true);
+    
+    try {
+      const forms = await metaformsApi.listMetaforms({});
+      const rows: Row[] = await Promise.all(forms.map(async form => {
+        return {
+          metaform: form,
+          versions: await versionsApi.listMetaformVersions({ metaformId: form.id!! })
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:52]:
```suggestion
          versions: await versionsApi.listMetaformVersions({ metaformId: form.id! })
```

File: src/components/screens/editor/editor-screen.tsx [hunk 19]
```diff
@@ -1,12 +1,273 @@
...
-  <Typography>{ strings.generic.notImplemented }</Typography>
-);
+const EditorScreen: React.FC = () => {
+  const errorContext = useContext(ErrorContext);
+  const navigate = useNavigate();
+
+  const apiClient = useApiClient(Api.getApiClient);
+  const { metaformsApi, versionsApi } = apiClient;
+
+  const [ metaforms, setMetaforms ] = useState<Row[]>([]);
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:36]: Maybe better describing would be metaformRows

**Review by EVP1989 — COMMENTED** [review]
Added some comments

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+/**
+ * Interface for Metaforms settings
+ */
+interface FormSettings {
+  formId: string;
+  formName: string;
+  formUrl: string;
+  formTemplate: boolean;
+  formSchema: string;
+  formIdentification: boolean;
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:26]: Done (formAuthentication)

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+/**
+ * Editor Screen Drawer component 
+ */
+const EditorScreenDrawer: React.FC<Props> = ({
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:32]: Done.

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+   * Handles Save icon click
+   */
+  const handleFormSubmit = () => {
+    createMetaform({
+      allowAnonymous: !formSettings.formIdentification,
+      title: formSettings.formName
+    });
+  };
+
+  /* eslint-disable no-unneeded-ternary */
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:64]: Removed as un-necessary.

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+  /**
+   * Event handler for input element change
+   * 
+   * @param event event
+   */
+  const onInputFieldChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { target: { name, value } } = event;
+    const updatedForm: any = { ...formSettings };
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:72]: Done!

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+
+  /**
+   * Renders Drawer info section
+   */
+  const renderDrawerInfoSection = () => {
+    return (
+      <Box
+        sx={{ padding: 2 }}
+      >
+        <Stack spacing={2}>
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:144]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+
+  /**
+   * Renders Drawer template section
+   */
+  const renderDrawerTemplateSection = () => {
+    return (
+      <Box
+        sx={{ padding: 2 }}
+      >
+        <Stack spacing={2}>
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:174]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 3 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:207]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 2 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:212]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 4 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:234]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 5 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:240]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 6 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:253]: Done

File: src/components/screens/editor/editor-screen.tsx (hunk 19 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:36]: Done.

File: src/components/screens/editor/editor-screen.tsx (hunk 18 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:52]: Done.

File: src/components/screens/editor/editor-screen.tsx (hunk 17 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:59]: Done

File: src/localization/fi.json (hunk 7 above)
Comment (villejuutila) [reply src/localization/fi.json:261]: Done

File: src/localization/en.json (hunk 8 above)
Comment (villejuutila) [reply src/localization/en.json:262]: Done

File: src/components/screens/editor/editor-screen.tsx (hunk 10 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:98]: Ah good point, thanks... :)

File: src/components/screens/editor/editor-screen.tsx (hunk 9 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:176]: Done!

File: src/components/screens/editor/editor-screen.tsx (hunk 11 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx]: Done

File: src/components/screens/editor/editor-screen.tsx (hunk 13 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx]: Done

File: src/components/screens/editor/editor-screen.tsx (hunk 12 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx]: Done

File: src/components/screens/editor/editor-screen.tsx (hunk 14 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+   * @param event event
+   */
+  const onInputFieldChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { target: { name, value } } = event;
+    if (name === "formTemplate" || name === "formAuthentication") {
+      setFormSettings({ ...formSettings, [name]: value === "true" });
+    } else {
+      setFormSettings({ ...formSettings, [name]: value });
+    }
+    console.log(formSettings);
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:79]: Remove console.log()

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+
+  /**
+   * Renders Drawer header
+   */
+  const renderDrawerHeader = () => {
+    return (
+      <Stack>
+        <Box sx={{
+          padding: 2, height: 72, backgroundColor: "#F5F5F5", display: "flex", flexDirection: "row", justifyContent: "space-between"
+        }}
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:97]:
```suggestion
        <Box sx={{
          padding: 2,
          height: 72,
          backgroundColor: "#F5F5F5",
          display: "flex",
          flexDirection: "row",
          justifyContent: "space-between"
        }}
```

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+              }}
+              onClick={ handleCloseClick }
+            >
+              <Clear color="primary"/>
+            </IconButton>
+          </Box>
+        </Box>
+        <Box
+          sx={{ height: 42, justifyContent: "center" }}
+        >
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:128]:
```suggestion
        <Box sx={{ height: 42, justifyContent: "center" }}>
```

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+  };
+
+  /**
+   * Renders Drawer info section
+   */
+  const renderDrawerInfoSection = () => {
+    return (
+      <Box
+        sx={{ padding: 2 }}
+      >
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:144]:
```suggestion
      <Box sx={{ padding: 2 }}>
```

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+  };
+
+  /**
+   * Renders Drawer authentication section
+   */
+  const renderDrawerAuthenticationSection = () => {
+    return (
+      <Box
+        sx={{ padding: 2 }}
+      >
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:212]:
```suggestion
      <Box sx={{ padding: 2 }}>
```

File: src/components/screens/editor/editor-screen-drawer.tsx [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+  useEffect(() => {
+    resetFormSchema();
+  }, [formSettings.formTemplate]);
+
+  return (
+    <Drawer
+      PaperProps={{
+        sx: {
+          width: "400px", borderRadius: "15px 0 0 15px"
+        }
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen-drawer.tsx:263]:
```suggestion
        sx: {
          width: "400px",
          borderRadius: "15px 0 0 15px"
        }
```

File: src/components/screens/editor/editor-screen.tsx [hunk 32]
```diff
@@ -1,12 +1,293 @@
...
+              }}
+            >
+              <AccordionSummary
+                sx={{
+                  backgroundColor: "white",
+                  padding: 0
+                }}
+                expandIcon={ <KeyboardArrowDown style={{ fill: "darkgrey" }}/> }
+              >
+                <AdminFormTypographyField variant="h3">{ params.row.title ?? "No title on Metaform" }</AdminFormTypographyField>
```
Comment (EVP1989) [inline src/components/screens/editor/editor-screen.tsx:200]: Metaform should always have title, it's required field.

**Review by EVP1989 — COMMENTED** [review]
Approved with comments

File: src/components/screens/editor/editor-screen.tsx [hunk 33]
```diff
@@ -1,12 +1,293 @@
...
+              <Grid item md={ 2 }>
+                <AdminFormTypographyField>
+                  <DateRangeRounded style={{ fill: "darkgrey", marginRight: "0.5rem" }}/>
+                  { moment(version.createdAt).format("DD.MM.yyyy HH:mm") }
+                </AdminFormTypographyField>
+              </Grid>
+              <Grid item md={ 2 }>
+                <AdminFormTypographyField>
+                  <DateRangeRounded style={{ fill: "darkgrey", marginRight: "0.5rem" }}/>
+                  { moment(version.modifiedAt).format("DD.MM.yyyy HH:mm") }
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:122]: Done.

File: src/components/screens/editor/editor-screen.tsx (hunk 32 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen.tsx:200]: Sounds reasonably. At some point there was a Metaform without a title but it seems to be changed by now.
Fixed.

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 26 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:79]: ...... Done..... :|

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 27 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:97]: Done.

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 28 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:128]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 29 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:144]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 30 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:212]: Done

File: src/components/screens/editor/editor-screen-drawer.tsx (hunk 31 above)
Comment (villejuutila) [reply src/components/screens/editor/editor-screen-drawer.tsx:263]: Done :)

**Review by EVP1989 — APPROVED** [review]
Approved

**Pull request merged**